examples/writing_sparse_heter_dim_test.go

```diff
@@ -0,0 +1,172 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * When run, this program will create a simple 2D sparse array, having
+ * heterogeneous dimensions and writes data to it
+ *
+ */
+
+package examples
+
+import (
+	"fmt"
+	"os"
+
+	tiledb "github.com/TileDB-Inc/TileDB-Go"
+)
+
+// Name of array.
+var heterArrayName = "writing_sparse_heter_dim"
+
+func createSparseHeterDimArray() {
+	// Create a TileDB context.
+	ctx, err := tiledb.NewContext(nil)
+	checkError(err)
+
+	// The array will be 4x4 with dimensions "rows" and "cols",
+	// with domain [1,4].
+	domain, err := tiledb.NewDomain(ctx)
+	checkError(err)
+	d1, err := tiledb.NewDimension(ctx, "d1", []float32{1.0, 20.0}, float32(5.0))
+	checkError(err)
+	d2, err := tiledb.NewDimension(ctx, "d2", []int64{1, 30}, int64(5))
+	checkError(err)
+	err = domain.AddDimensions(d1, d2)
+	checkError(err)
+
+	// The array will be sparse.
+	schema, err := tiledb.NewArraySchema(ctx, tiledb.TILEDB_SPARSE)
+	checkError(err)
+	err = schema.SetDomain(domain)
+	checkError(err)
+
+	// Add a single attribute "a" so each (i,j) cell can store an integer.
+	a, err := tiledb.NewAttribute(ctx, "a", tiledb.TILEDB_INT32)
+	checkError(err)
+	err = schema.AddAttributes(a)
+	checkError(err)
+
+	// Create the (empty) array on disk.
+	array, err := tiledb.NewArray(ctx, heterArrayName)
+	checkError(err)
+	err = array.Create(schema)
+	checkError(err)
+}
+
+func writeSparseHeterDimArray() {
+	ctx, err := tiledb.NewContext(nil)
+	checkError(err)
+
+	// Write some simple data to cells.
+	buffD1 := []float32{1.1, 1.2, 1.3, 1.4}
+	buffD2 := []int64{1, 2, 3, 4}
+	buffA := []int32{1, 2, 3, 4}
+
+	// Open the array for writing and create the query.
+	array, err := tiledb.NewArray(ctx, heterArrayName)
+	checkError(err)
+	err = array.Open(tiledb.TILEDB_WRITE)
+	checkError(err)
+	query, err := tiledb.NewQuery(ctx, array)
+	checkError(err)
+	err = query.SetLayout(tiledb.TILEDB_UNORDERED)
+	checkError(err)
+	_, err = query.SetBuffer("d1", buffD1)
+	checkError(err)
+	_, err = query.SetBuffer("d2", buffD2)
+	checkError(err)
+	_, err = query.SetBuffer("a", buffA)
+	checkError(err)
+
+	// Perform the write and close the array.
+	err = query.Submit()
+	checkError(err)
+	err = array.Close()
+	checkError(err)
+}
+
+func readSparseHeterDimArray() {
+	ctx, err := tiledb.NewContext(nil)
+	checkError(err)
+
+	// Prepare the array for reading
+	array, err := tiledb.NewArray(ctx, heterArrayName)
+	checkError(err)
+	err = array.Open(tiledb.TILEDB_READ)
+	checkError(err)
+
+	buffD1R := make([]float32, 4)
+	buffD2R := make([]int64, 4)
+	buffAR := make([]int32, 4)
+
+	// Prepare the query
+	query, err := tiledb.NewQuery(ctx, array)
```

Comment:
For completeness can you add a a check for the nonEmptyDomain too?

Reply:
Sure, done